xds/src/main/java/io/grpc/xds/FaultFilter.java

```diff
@@ -215,23 +218,48 @@ public ClientInterceptor buildClientInterceptor(
           // TODO(https://github.com/grpc/grpc-java/issues/7868)
           callExecutor = MoreExecutors.directExecutor();
         }
-        if (finalDelayNanos != null && finalAbortStatus != null) {
-          return new DelayInjectedCall<>(
-              finalDelayNanos, callExecutor, scheduler, callOptions.getDeadline(),
-              Suppliers.ofInstance(
-                  new FailingClientCall<ReqT, RespT>(finalAbortStatus, callExecutor)));
-        }
-        if (finalAbortStatus != null) {
-          return new FailingClientCall<>(finalAbortStatus, callExecutor);
+        if (finalDelayNanos != null) {
+          Supplier<? extends ClientCall<ReqT, RespT>> callSupplier;
+          if (finalAbortStatus != null) {
+            callSupplier = Suppliers.ofInstance(
+                new FailingClientCall<ReqT, RespT>(finalAbortStatus, callExecutor));
+          } else {
+            callSupplier = new Supplier<ClientCall<ReqT, RespT>>() {
+              @Override
+              public ClientCall<ReqT, RespT> get() {
+                return next.newCall(method, callOptions);
+              }
+            };
+          }
+          final DelayInjectedCall<ReqT, RespT> delayInjectedCall = new DelayInjectedCall<>(
+              finalDelayNanos, callExecutor, scheduler, callOptions.getDeadline(), callSupplier);
+
+          final class DeadlineInsightForwardingCall extends ForwardingClientCall<ReqT, RespT> {
+            @Override
+            protected ClientCall<ReqT, RespT> delegate() {
+              return delayInjectedCall;
+            }
+
+            @Override
+            public void start(Listener<RespT> listener, Metadata headers) {
+              Listener<RespT> finalListener =
+                  new SimpleForwardingClientCallListener<RespT>(listener) {
+                    @Override
+                    public void onClose(Status status, Metadata trailers) {
+                      if (status.getCode().equals(Code.DEADLINE_EXCEEDED)) {
```

Comment:
Is this condition sufficient? @ejona86 
```java
status.getCode().equals(Code.DEADLINE_EXCEEDED)
```

Reply:
Uhh... this entire approach seems problematic. We don't modify statuses sent by a server; we only modify statuses that we generate locally. So doing this would break a long-time invariant. I'm not seeing an obvious solution.

Reply:
I do see we modify status sent by server already:
https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ClientCallImpl.java#L692-L701

Reply:
That is _replacement_ not modification. And it is safe to do because there were two errors: the one we got from the remote and the one that happened locally. We simply chose to use the local one in that case. Note that it fully-replaces the status and metadata, so it is like we never received the trailers from the server.

Reply:
Recreated status and metadata. As discussed offline, ideally we should only replace status and metadata if locally deadline is exceeded. However, it's not easy to get the effective locally deadline in an interceptor (A StreamTracer might be needed), and it might not worth that amount of effort just for adding debug information. Given that, reporting a delay injected hint in error message is still better than nothing.